internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,346 @@
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		return nil, status.Error(codes.InvalidArgument,
+			fmt.Sprintf("targetpath %v is empty", targetPath))
+	}
+
+	stat, err := os.Stat(targetPath)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"failed to get stat for targetpath %q: %v", targetPath, err)
+	}
+
+	if stat.Mode().IsDir() {
+		return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath)
+	}
+
+	return nil, status.Errorf(codes.InvalidArgument,
+		"targetpath %q is not a directory or device", targetPath)
+}
+
+// mountNFS mounts nfs volumes.
+func (ns *NodeServer) mountNFS(
+	ctx context.Context,
+	volumeID, clusterID, source, mountPoint string,
+	mountOptions []string,
+	mountPermissions uint64,
+) error {
+	var (
+		stderr, netNamespaceFilePath string
+		err                          error
+	)
+
+	if clusterID != "" {
+		netNamespaceFilePath, err = util.GetNFSNetNamespaceFilePath(
+			util.CsiConfigFile,
+			clusterID)
+		if err != nil {
+			return err
+		}
+	}
+
+	notMnt, err := ns.Mounter.IsLikelyNotMountPoint(mountPoint)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(mountPoint, os.FileMode(mountPermissions)); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+	if !notMnt {
+		log.DebugLog(ctx, "nfs: volume is already mounted to %s", mountPoint)
+
+		return nil
+	}
+
+	args := []string{
+		"-t", "nfs",
+		source,
+		mountPoint,
+	}
+
+	if len(mountOptions) > 0 {
+		args = append(append(args, "-o"), mountOptions...)
+	}
+
+	log.DefaultLog("nfs: mounting volumeID(%v) source(%s) targetPath(%s) mountflags(%v)",
+		volumeID, source, mountPoint, mountOptions)
+	if netNamespaceFilePath != "" {
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
+	}
+	if err != nil {
+		return fmt.Errorf("nfs: failed to mount %q to %q : %w stderr: %q",
+			source, mountPoint, err, stderr)
+	}
+	if stderr != "" {
+		return fmt.Errorf("nfs: failed to mount %q to %q : stderr %q",
+			source, mountPoint, stderr)
+	}
+
+	return err
+}
+
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
+	switch {
+	case req.GetVolumeId() == "":
+		return errors.New("volume ID missing in request")
+	case req.GetVolumeCapability() == nil:
+		return errors.New("volume capability missing in request")
+	case req.GetTargetPath() == "":
+		return errors.New("target path missing in request")
+	case req.GetVolumeContext()[paramServer] == "":
+		return fmt.Errorf("%v missing in request", paramServer)
+	case req.GetVolumeContext()[paramShare] == "":
+		return fmt.Errorf("%v missing in request", paramShare)
+	}
+
+	return nil
+}
+
+// chmodIfPermissionMismatch only performs chmod when permission mismatches.
```

Comment:
This isn't recursive, only for a single directory/file. That probably needs explaining. A unit-test would be good to have too. On the other hand, if fsGroupPolicy or similar can cover this, that would be much better.